app/src/main/java/com/keylesspalace/tusky/components/compose/ComposeActivity.kt

```diff
@@ -899,25 +919,57 @@ class ComposeActivity :
         val context = getApplicationContext()
-        val tempFile = createNewImageFile(context, if (isPng) ".png" else ".jpg")
+        val tempFile = createNewImageFile(context,
+            if (isPng) { ".png" }
+            else if (isImage) { ".jpg" }
+            else {
+                val extension = if (mimeType != null) MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType) else null
+                if (extension != null) "." + extension
+                else "" // FIXME: Would be better to error? Without a mime type the intent won't form right
+            }
+        )
 
-        // "Authority" must be the same as the android:authorities string in AndroidManifest.xml
+        // "Authority" here must be the same as the android:authorities string in AndroidManifest.xml
         val uriNew = FileProvider.getUriForFile(context, BuildConfig.APPLICATION_ID + ".fileprovider", tempFile)
 
-        viewModel.cropImageItemOld = item
+        val editInternally = isImage
+        if (editInternally) { // We can use Android-Image-Cropper
+            viewModel.cropImageItemOld = item
 
-        cropImage.launch(
-            options(uri = item.uri) {
-                setOutputUri(uriNew)
-                setOutputCompressFormat(if (isPng) Bitmap.CompressFormat.PNG else Bitmap.CompressFormat.JPEG)
+            cropImage.launch(
+                options(uri = item.uri) {
+                    setOutputUri(uriNew)
+                    setOutputCompressFormat(if (isPng) Bitmap.CompressFormat.PNG else Bitmap.CompressFormat.JPEG)
+                }
+            )
+        } else { // Pawn off to a external editor
+            val intent = Intent(Intent.ACTION_EDIT)
+            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
+            intent.setDataAndType(item.uri, mimeType)         // Where will media be loaded from?
+            intent.putExtra(MediaStore.EXTRA_OUTPUT, uriNew) // Where will media be saved to?
+
+            // FIXME: This grants more permissions to more apps than are necessary
+            val resInfoList = context.getPackageManager().queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
+            for (resolveInfo in resInfoList) {
+                val packageName = resolveInfo.activityInfo.packageName;
+                context.grantUriPermission(packageName, uriNew, Intent.FLAG_GRANT_WRITE_URI_PERMISSION or Intent.FLAG_GRANT_READ_URI_PERMISSION);
             }
-        )
+            try {
+                startActivityForResult(intent, MEDIA_EDIT_REQUESTCODE) // FIXME: Would using Intent.createChooser(intent) be better?
+            } catch (e: ActivityNotFoundException) {
+                // FIXME: Would be nice if error message could be specialized on (or flat out include) the mime type
+                displayTransientError(R.string.error_media_edit_intent_failed)
+            } 
```

Comment:
Ideally, this code would use a contract, not `startActivityForResult`:
* startActivityForResult is actually deprecated. They want you to use contracts.
* using startActivityForResult requires overriding onActivityResult, which is a little unpleasant (and is also deprecated)
* startActivityForResult requires some very bad permissions code (see FIXME above). Basically, when using startActivity any URL the target activity is going to edit must be specifically granted for its use with `grantUriPermission`. **But**, since we are using an implicit intent, we do not know ahead of time what activity is going to handle the url! So the code here (adapted from StackOverflow) grants edit permissions to *every program on the system* that has the ability to edit URLs of the appropriate MIME type *at all*, just in case it is the one the user chooses. This is not as secure as it should be. Comments on StackOverflow suggest that using Intent.createChooser() might magically wire the permissions up correctly, but the correct way to do this is with a contract. 

Unfortunately, there is no [standard contract](https://developer.android.com/reference/androidx/activity/result/contract/ActivityResultContracts) that handles the case we are looking for (`android.intent.action.EDIT`). There is a `ActivityResultContracts.TakePicture` that will take a picture and write it to a URL, but we need something that accepts one input URL and also writes to one output URL. We would need to create a custom contract. I do not know where to start with that :(

Reply:
Check `com.keylesspalace.tusky.util.PickMediaFiles`, it is a custom contract for the picker, and make a similar one for media editing